File: 643_maximum_avg_subarray_i/sliding_window.py
class Solution:
    def findMaxAverage(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: float
        """
        """
        Complexity: O(n)
        Space: O(1)
        """
        tpsum = sum(nums[0:k])
        maxsum = tpsum
        for i in range(1, len(nums)-k + 1):
            # if directly compute the sum of each posibilities of subarrays
            # the complexity becomes O(n*k)
            # it would lead to huge computation time when k ~ n
            tpsum = tpsum - nums[i-1] + nums[i+k-1]
            if tpsum > maxsum:
                maxsum = tpsum
        return (1.0 * maxsum) / k

File: 643_maximum_avg_subarray_i/sliding_window.cpp
class Solution {
  /*
   * Complexity: O(n)
   * Space: O(1)
   */
public:
    double findMaxAverage(vector<int>& nums, int k) {
        int n = nums.size();
        int max = 0;
        
        for(int i = 0; i < k; ++i){
            max += nums[i];
        }
        
        int temp_max = max;
        for(int i = k; i < n ; ++i){
            temp_max -= nums[i - k];
            temp_max += nums[i];
            max = (temp_max > max) ? temp_max : max;
        }
        
        return (max*1.0) / k;
    }
};
